letsencrypt-aws.py

```diff
@@ -1,154 +1,266 @@
-            LoadBalancerName=self.elb_name,
-            SSLCertificateId=new_cert_arn,
-            LoadBalancerPort=self.elb_port,
-        )
+
+def _get_cert_arn(cert):
+    return cert.server_certificate_metadata['Arn']
+
+
+def _write_cert(cert):
+    with open(cert.name + '.key', 'w') as key:
+        key.write(cert.key)
+    with open(cert.name + '.crt', 'w') as crt:
+        crt.write(cert.cert_body)
+    with open(cert.name + '.chain', 'w') as chain:
+        chain.write(cert.chain)
+
+
+def _read_cert_from_disk(name):
+    with open(name + '.key') as key:
+        private_key = key.read()
+    with open(name + '.crt') as crt:
+        cert_body = crt.read()
+    with open(name + '.chain') as chain:
+        cert_chain = chain.read()
+    return Certificate(name, private_key, cert_body, cert_chain)
+
+
+def create(acme_client,
+           route53_client,
+           hosts,
+           path,
+           name,
+           save_local_copy=False):
+    iam = boto3.resource('iam')
+    challenge_completer = Route53ChallengeCompleter(route53_client)
+    logger.info('Creating certificate: {}'.format(name))
+    new_cert = _provision_cert(
+        name, hosts, acme_client, challenge_completer)
+    if(save_local_copy):
+        logger.info('Saving certificate {} to disk'.format(name))
+        _write_cert(new_cert)
+    logger.info('Uploading new {} certificate'.format(name))
+    _upload_cert(new_cert, iam, path)
+
+
+def update(acme_client,
+           route53_client,
+           path,
+           expiration_threshold,
+           regions,
+           read_local_copy=False,
+           save_local_copy=False):
+    iam = boto3.resource('iam')
+    challenge_completer = Route53ChallengeCompleter(route53_client)
+    for cert in _get_expiring_certificates(iam, path, expiration_threshold):
+        if read_local_copy:
+            logger.info('Reading cert {} from disk'.format(cert.name))
+            new_cert = _read_cert_from_disk(cert.name)
+        else:
+            logger.info('Renewing certificate: {}'.format(cert.name))
+            hosts = _get_domains_for_cert(cert.certificate_body)
+            new_cert = _provision_cert(
+                cert.name, hosts, acme_client, challenge_completer)
+        if(save_local_copy):
+            logger.info('Saving certificate {} to disk'.format(cert.name))
+            _write_cert(new_cert)
+        logger.info('Renaming {} certificate'.format(cert.name))
+        old_cert = cert.update(NewServerCertificateName=cert.name + '-old')
+        logger.info('Uploading renewed {} certificate'.format(cert.name))
+        renewed_cert = _upload_cert(new_cert, iam, path)
+        logger.info('Finding ELBs with cert {}'.format(old_cert))
+        _change_elb_cert(old_cert, renewed_cert, regions)
+        logger.info('Deleting old certificate: {}'.format(old_cert.name))
+        old_cert.delete()
+        logger.info('Done renewing certificate: {}'.format(cert.name))
```

Comment:
What happens when this renewal logic of write cert, rename old, upload new, switch elb to use new, delete old fails?  Is the ELB or old cert left in a weird state?  Is there enough info in logs to fix the issue?  Should it attempt retries and/or otherwise try to be somewhat transactional?